lib/ClangImporter/ClangImporter.cpp

```diff
@@ -3832,8 +3837,23 @@ void ClangImporter::Implementation::lookupValue(
     ValueDecl *decl;
     // If it's a Clang declaration, try to import it.
     if (auto clangDecl = entry.dyn_cast<clang::NamedDecl *>()) {
-      decl = cast_or_null<ValueDecl>(
-          importDeclReal(clangDecl->getMostRecentDecl(), CurrentVersion));
+      // When looking for a namespace, we need to populate all of its members.
+      // To do this, let's import all of its redeclarations.
```

Comment:
I'm trying to decide if this is the right place for this logic to exist. I think maybe it's not. 

Why did you decide to do this here rather than in namespace's visitor? Is there some problem with the extensions being added to the wrong module if you do it there?

Reply:
I think the problem with doing it in `VisitNamespaceDecl` is mostly that we're importing stuff that we weren't asked to, like redecls from other modules (and their contents). For example, if we're processing a SourceKit request to produce a module interface for C++ module `A` containing `namespace base`, IIUC we'd fail if another C++ module `B` has an invalid decl in `namespace base`.

On the other hand, when compiling Swift code we'd really like to have access to all the imported C++ code, so loading all the redecls looks okay to me.

You're right about extensions being added to wrong modules – when importing a namespace decl, we need the context of a specific redeclaration, not the canonical decl. We could solve this by putting all the code in `VisitNamespaceDecl` under `for (auto ns : decl->redecls)`, but I'm not sure that would be any better than the current approach.

Reply:
> For example, if we're processing a SourceKit request to produce a module interface for C++ module A containing namespace base, IIUC we'd fail if another C++ module B has an invalid decl in namespace base.

Sourcekit doesn't use `lookupValue`? That very well may be the case, I just don't know, and it would be a good thing to find out. 

Anyway, thanks for explaining. I pulled this code down and played with it a bit. I think this way of doing it makes sense. I just want to ensure that we're getting the benefit of moving the logic to `lookupValue` (enough benefit that it justifies the added complexity).

Can you please add two comments here (or one comment saying two things, whichever you think is best): first, one explaining why we look at all the redecls here and not in the visitor (as I understand it, this is to make sourcekit faster). Second, a comment saying that redecls can live across multiple modules.

Reply:
Hmm, on a closer look SourceKit actually does use this logic. So my argument about SourceKit performance doesn't hold up.

However, I've tried moving this logic to the visitor, and uncovered another problem. It has to do with the fact that `VisitNamespaceDecl` sometimes itself recursively, and those calls might call `VisitNamespaceDecl` again for a different redecl of the same declaration.

1. Let's say we're importing `typedef basic_string<...> string` declared inside `std::__1`.
2. For that we need to import `basic_string`, located in `std::_1` as well, but in a different redecl of `std::__1`.
3. Now we've ended up with two calls to `VisitNamespaceDecl` for different redecls of `std::__1` in the same call stack, and each of these calls will attempt to import all the other redecls as well.

This leads to parts of C++ classes not being properly imported. With the string example, `string::size_type` & `string::difference_type` aren't going to be available in Swift (because they require `std::__1::allocator_traits` located in yet another redecl of `std::_1`).



I added a comment with the brief reasoning, let me know if you'd like me to add some more details.

Reply:
Hmm, I ran into that problem as well when I was originally implementing this. 

In other parts of the importer, we resolve this by caching the namespace before importing any members (or in this case, before we look through the redecls) then we let the `importDecl` caching logic do the rest. Did you make sure that was implemented properly (i.e., cache before looking through the redecls, and make sure to use the cached version if it exists before getting to a recursive state)?

If there's some reason that doesn't work here that I'm forgetting, let me know. Otherwise, let's see if that works. (Or ping me if you want to discuss further).

Reply:
@egorzhdan any updates on this?

Reply:
Not yet, sorry for that :(

So far I wasn't able workaround the issue with `std::string::size_type` not being available when importing all the redecls inside `VisitNamespaceDecl`. I tracked down the cause of it, turns out we're importing `size_type` when processing an out-of-line decl for `std::string::npos`:
```cpp
template<class _CharT, class _Traits, class _Allocator>
_LIBCPP_FUNC_VIS
const typename basic_string<_CharT, _Traits, _Allocator>::size_type
               basic_string<_CharT, _Traits, _Allocator>::npos;
```

And then, when importing the actual struct members, we're skipping `size_type` here:

https://github.com/apple/swift/blob/9f5cf973c3f2b92aa7827fd2a7cc4372e817f498/lib/ClangImporter/ImportDecl.cpp#L3551-L3555

So we might actually be able to fix this specific issue with a completely different patch by skipping the already imported members less aggressively. I haven't made any progress on that yet, though.

Reply:
(Submitted a separate patch https://github.com/apple/swift/pull/38439 to fix the issue described in the above comment)

Reply:
So, if we apply both this patch and #38439, then we can move this logic back to `VisitNamespaceDecl`?

Reply:
Yeap, I moved it back to the visitor.